From b43c46a2db5ae02ba77edd077bb685f3143fa884 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Fri, 25 Apr 2025 12:19:13 +0200
Subject: [PATCH] clk: stm32mp1: don't disable a PLL if DIV P/Q/R are ON

Don't disable a PLL if DIVPEN, DIVQEN, DIVREN are set to ON.

Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: Ieccd90ff571c62bba71e59f7e747fdfa72fbc7fe
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/454382
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/clk/stm32/clk-stm32mp1.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

--- a/drivers/clk/stm32/clk-stm32mp1.c
+++ b/drivers/clk/stm32/clk-stm32mp1.c
@@ -913,6 +913,8 @@ static const struct clk_ops clk_pll_frac
 #define PLL_FRAC_SHIFT		3
 #define PLL_FRAC_WIDTH		13
 
+#define PLL_DIV_PQR_MASK	GENMASK(6, 4)
+
 #define TIMEOUT 5
 
 static int pll_enable(struct clk_hw *hw)
@@ -939,10 +941,14 @@ static int pll_enable(struct clk_hw *hw)
 
 static void pll_disable(struct clk_hw *hw)
 {
+	struct clk_gate *gate = to_clk_gate(hw);
+
 	if (!clk_gate_ops.is_enabled(hw))
 		return;
 
-	clk_gate_ops.disable(hw);
+	/* Disable only if DIV P/N/Q are not set to ON */
+	if (!(readl_relaxed(gate->reg) & PLL_DIV_PQR_MASK))
+		clk_gate_ops.disable(hw);
 }
 
 const struct clk_ops pll_gate_ops = {
